src/runtime/execution.rs

```diff
@@ -484,7 +491,9 @@ impl ExecutionState {
             .map(|t| t.id)
             .collect::<SmallVec<[_; DEFAULT_INLINE_TASKS]>>();
 
-        if runnable.is_empty() {
+        let any_blocked = self.tasks.iter().any(|t| t.blocked() && !t.dropped);
+
+        if runnable.is_empty() || (!any_blocked && runnable.iter().all(|id| self.get(*id).dropped)) {
```

Comment:
I think this logic deserves a comment, it's a bit tricky to follow the cases that count as "Finished" now.

Also, to avoid iterating over `self.tasks` twice, we could probably do something like
```
let mut blocked_attached = false;
let runnable = self
    .tasks
    .iter()
    .inspect(|t| blocked_attached = blocked_attached || (t.blocked() && !t.dropped))
    .filter(|t| t.runnable())
    .map(|t| t.id)
    .collect::<SmallVec<[_; DEFAULT_INLINE_TASKS]>>();
```

Reply:
Yeah it's a bit tricky. I'll add a comment.